### PR TITLE
react/week4

### PR DESCRIPTION
**Pull request by amerdahabi**

### CODE REVIEW

File: react/week4/my-app/src/users.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  
+    if (error) {
+      return <div>{`error fetching ${error}...`}</div>;
+    }
+  
+    if (loading) {
+      return <div>Loading...</div>;
+    }
+  
+    if (!users || users.length === 0) {
```
Comment (jorgecarleitao) [inline react/week4/my-app/src/users.js:16]: Is there any situation where `!users` is needed? I think that users is always a list. If it is not, we should probably try to validate the result of the API call to make it so.

In general, we strive to keep the state (in a component, context or redux) of the same type throughout the application.

File: react/week4/my-app/src/users.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      return <div>No result..</div>;
+    }
+  
+    return <ul>
+        {users.map(user => {
```
Comment (jorgecarleitao) [inline react/week4/my-app/src/users.js:21]: when a anonymous function has a single return, the code can be even more simplified:

```user => <li key={user.id}>{user.login}</li>``` 

(no curly brackets and no return)

File: react/week4/my-app/src/user.context.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    return (
+      <UserContext.Provider
+        value={{ ...this.state, updateUserContext: this.updateUserContext }}
+      >
+        {this.props.children}
+      </UserContext.Provider>
+    );
+  }
+}
+ export default UserProvider
```
Comment (jorgecarleitao) [inline react/week4/my-app/src/user.context.js:26]: Note that there is no end of line here (`\n`).

In general, we try to keep an end of line character on every line, to indicate that the line ended.

Another advantage of having an end of line is the following: suppose the following code:

```
a = 1[\n]
b = 2
```
When we want to add a new line to this code, 

```
a = 1[\n]
b = 2[\n]
c = 3[\n]
```

we needed to add an end of line in line `b = 2`. This implies that git will tell us that there was one line changed (`b = 2` -> `b = 2[\n]` and one new line `c = 3[\n]`. However, this does not represent well what we were really doing: we were just adding a new line.

**Review by jorgecarleitao — APPROVED** [review]
Really well done: great code and semantic. Congrats!

I left some minor comments. They are more "good to know" for you.

File: react/week4/my-app/src/user.context.js (hunk 3 above)
Comment (amerdahabi) [reply react/week4/my-app/src/user.context.js:26]: HI Jorge, thank you so much for your feed back 😊.
yes that make sense of course, thanks for explain that

File: react/week4/my-app/src/users.js (hunk 1 above)
Comment (amerdahabi) [reply react/week4/my-app/src/users.js:16]: Actually before i added (!users) when i typing in the input, it all works good, but when i delete the content of the input i got this error (TypeError: Cannot read property 'length' of undefined).

File: react/week4/my-app/src/users.js (hunk 2 above)
Comment (amerdahabi) [reply react/week4/my-app/src/users.js:21]: thanks for mentioning this 👍🏻

File: react/week4/my-app/src/users.js (hunk 1 above)
Comment (jorgecarleitao) [reply react/week4/my-app/src/users.js:16]: I imagined that there was a good reason for it to be there :)

Ideally, I would track it down fix it (wherever it is being assigned no `undefined`). It is fine that you do not do it for the homework, but keep the idea in mind: strive to keep the state of the same type throughout the lifetime of the application. Only use `undefined` or `null` when it represents a specific situation (e.g. user id for an anonymous user is often `undefined` or `null).

**Pull request merged**